fdk.go

```diff
@@ -143,52 +142,41 @@ type callResponseHTTP struct {
 }
 
 type jsonOut struct {
-	jsonio
-	Protocol *callResponseHTTP `json:"protocol,omitempty"`
-
-	io.Writer
+	Body        string           `json:"body"`
+	ContentType string           `json:"content_type"`
+	Protocol    callResponseHTTP `json:"protocol,omitempty"`
 }
 
-func (out *jsonOut) Write(p []byte) (n int, err error) {
-	out.Body += string(p)
-	return len(p), nil
+func (out *jsonOut) StatusCode() int {
+	return out.Protocol.StatusCode
 }
 
 func (out *jsonOut) WriteStatus(status int) {
 	out.Protocol.StatusCode = status
 }
 
-func doJSONOnce(handler Handler, ctx context.Context, in io.Reader, out io.Writer, buf *bytes.Buffer, hdr http.Header) {
+func doJSONOnce(handler Handler, ctx context.Context, in io.Reader, out io.Writer, jsonResponse *jsonOut, jsonRequest *jsonIn, buf *bytes.Buffer, hdr http.Header) {
 	buf.Reset()
 	resetHeaders(hdr)
 
-	resp := &jsonOut{
-		jsonio: jsonio{
-			ContentType: "application/json",
-		},
-		Protocol: &callResponseHTTP{
-			StatusCode: 200,
-		},
-		Writer: buf,
-	}
-
-	req := &jsonIn{}
-	err := json.NewDecoder(in).Decode(req)
+	err := json.NewDecoder(in).Decode(jsonRequest)
 	if err != nil {
-		resp.WriteStatus(500)
-		resp.Body = fmt.Sprintf(`{"error": %v}`, err.Error())
+		jsonResponse.WriteStatus(500)
+		jsonResponse.Body = fmt.Sprintf(`{"error": %v}`, err.Error())
 	} else {
-		setHeaders(ctx, req.Protocol.Headers)
+		setHeaders(ctx, jsonRequest.Protocol.Headers)
 		//TODO(xxx): use FN_DEADLINE here
 		ctx, cancel := context.WithCancel(ctx)
 		defer cancel()
-		handler.Serve(ctx, strings.NewReader(req.Body), resp)
-		resp.WriteStatus(200)
+		handler.Serve(ctx, strings.NewReader(jsonRequest.Body), buf)
+		jsonResponse.WriteStatus(200)
```

Comment:
2 things with `WriteStatus`:

* our JSON protocol allows a user to send back a custom response code to reply to the client with, here they can only send 200 if their function runs ok (regardless of if they want to return another code) since we aren't checking for the status that the user may have written, to allow that:
* since we're not sending `jsonResponse` to `handler.Serve` the user has no way to set a status code. the way for them to do that is https://github.com/fnproject/fdk-go/blob/master/fdk.go#L63 -- we can easily change this code to:

```go
cw, ok := out.(interface{
  WriteStatus(code int)
})
if ok {
  cw.WriteStatus(code)
}
```

and then this code will need to change to let `jsonResponse` or an object like it that can save a status implement `io.Reader` to return the body as well as allow status writing. open to ideas for better ways to allow `WriteStatus` for the client, what we have seems like we can hide the magic pretty well and make it pretty slick with the outlined methodology.

Reply:
better to reuse response object that is already there, so just needed to adjust json request handler